### ISSUE TITLE
Safari `readableStreamTransfer` workaround turns `null` body into empty `ArrayBuffer`, requests fail

### ISSUE THREAD

**Issue by david-crespo** [kettanaito/msw#1827]
### Prerequisites

- [X] I confirm my issue is not in the [opened issues](https://github.com/mswjs/msw/issues)
- [X] I confirm the [Frequently Asked Questions](https://mswjs.io/docs/faq) didn't contain the answer to my issue

### Environment check

- [X] I'm using the [latest](https://github.com/mswjs/msw/releases/latest) `msw` version
- [X] I'm using Node.js version 14 or higher

### Browsers

Safari

### Reproduction repository

https://github.com/oxidecomputer/msw2-safari-bug/

### Reproduction steps

`npm install`, `npm run dev`, open in safari, see error on page and in dev console.

### Current behavior

In Safari only, handler responses with `null` body produce the following error:

```
FetchEvent.respondWith received an error: TypeError: Response cannot have a body with the given status.
```

The error is thrown here on the `new Response(...)`

https://github.com/mswjs/msw/blob/008a8c69610709308fee5226e24f627717882bcb/src/mockServiceWorker.js#L275-L292

because `response.body` is an `ArrayBuffer` even though it was `null` in the handler definition. When the status is `204` (or other null body status) but the body is something other than `null`, the `Response` constructor throws an error:

<img width="734" alt="image" src="https://github.com/mswjs/msw/assets/3612203/bc4d950d-5eff-483d-a9b5-7ae22c3d7211">

### Suspected cause

I believe this is caused by this workaround for the lack of `readableStreamTransfer` in Safari:

https://github.com/mswjs/msw/blob/008a8c69610709308fee5226e24f627717882bcb/src/browser/setupWorker/start/createRequestListener.ts#L52-L73

Specifically, the line 

```ts
const responseBuffer = await responseClone.arrayBuffer()
```

will give a non-null empty `ArrayBuffer` even when `responseClone.body` is `null`. (The screenshot is in Firefox, but it's the same in Chrome and Safari.)

<img width="410" alt="image" src="https://github.com/mswjs/msw/assets/3612203/0e4133de-f819-49b5-841d-682a0c4b50f8">

I believe the fix (or _a_ fix) would be something like 

```ts
const responseBuffer = responseClone.body === null ? null : await responseClone.arrayBuffer()
```

I'm happy to make that PR, I just wanted to write this all down after a fun day of tracking it down with @zephraph.

### Expected behavior

`new Response(null, { status: 204 })` works in Safari like it does anywhere else.

**Comment by kettanaito** [comment]
Hey, @david-crespo. Thanks for reporting this. 

You are right about it. Also, welcome to JavaScript:

```js
let one = new Response(null, { status: 204 })
let two = new Response(await one.arrayBuffer(), { status: 204 })
// VM288:2 Uncaught TypeError: Failed to construct 'Response': Response with null body status cannot have body
```

I wish we wouldn't have to do these dances with `=== null` and checking for special no-body status codes.

**Comment by just-be-dev** [comment]
While we're fixing this, we should also expand the definition beyond just 204s. @david-crespo found the [code in webkit](https://opensource.apple.com/source/WebCore/WebCore-7603.3.8/Modules/fetch/FetchResponse.js.auto.html) that generates the resulting error. The check is...

```js
if (body !== @undefined && body !== null) {
        if (status === 101 || status === 204 || status === 205 || status === 304)
            @throwTypeError("Response cannot have a body with the given status");
```

So we should at least cover ~`101`,~ `204`, `205`, and `304` as we know those are specifically rejected by webkit. ~The [spec](https://fetch.spec.whatwg.org/#statuses) also includes `103` in that list, so it's probably best to just follow what the spec says.~

I tested 101 and 103 response codes across browsers and discovered they all fail with a range error where they expect a status code between 200 and 599. Given that, we can skip the 1XX codes. 

I'd also recommend we try to unify these checks as they're duplicated here:

https://github.com/mswjs/msw/blob/008a8c69610709308fee5226e24f627717882bcb/src/mockServiceWorker.js#L123-L128

**Comment by just-be-dev** [comment]
If others stumble upon this issue and have the need for an immediate workaround, we've mitigated the issue my updating `mockServiceWorker.js` directly: https://github.com/oxidecomputer/console/commit/7a930f20f0861f3f9abffc633010d1512b211b23.

**Comment by kettanaito** [comment]
@zephraph we don't really need to do any guarding on our end: MSW operates on the premise that the responses you construct are valid. JavaScript itself will validate that (it's _you_ who does `new Response(...)` in the handlers). No need to check for status codes.

On our end, this is an internal issue of treating response bodies a bit differently for Safari only since Safari has a bug of not being able to transfer `ReadableStream` over a `MessageChannel`. So, your initial proposal of `x.body === null ? null : await x.arrayBuffer()` would work nicely.

**Comment by kettanaito** [comment]
> I'd also recommend we try to unify these checks as they're duplicated here:

This is a noble suggestion but the worker script doesn't have a build step so reusing the same utilities will be problematic (workers have their own strict scope, so if you wish to have imports in a worker script, you have to then resolve (bundle) those imports on build time). Historically, I kept the worker script build-less so it's literally 1-1 to what the users would get in their public directories. I think it's a great advantage to have.

**Comment by kettanaito** [comment]
I've prepared the fix, should be published in the next patch release.

**Comment by kettanaito** [comment]
## Released: v2.0.3 🎉

This has been released in v2.0.3!

- 📄 [**Release notes**](https://github.com/mswjs/msw/releases/tag/v2.0.3)
- 📦 [npm package](https://www.npmjs.com/package/msw/v/2.0.3)

Make sure to always update to the latest version (`npm i msw@latest`) to get the newest features and bug fixes.

---

_Predictable release automation by [@ossjs/release](https://github.com/ossjs/release)_.
